### PR TITLE
fix: class field is not accessible via super

### LINKED ISSUE #35314: Accessing property on super should be a type error
**TypeScript Version:**  3.7.2

**Search Terms:**

typescript property access super undefined

**Code**

```ts
    class Base {
        n = 3
    }

    class Foo extends Base {
        test() {
            console.log(super.n) // type system says it should be a number, but it is undefined
        }
    }

    new Foo().test() // logs "undefined"
```

**Expected behavior:**

Should be some sort of type error, because clearly it is undefined

**Actual behavior:**

no error until runtime

**[Playground Link](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=2&pc=1#code/AQ4YwGwQwZx4BCsCmwDeBYAUKXwB2wAvMAMza4C+2Foks8AYgPbPDIAeALsvgCbwkMVJhx4QPGFwAUASnS1xuMM3wxmEZADoIzAObSYAVwAOyAE5b88gPQ3gXAJ5ngMR1OQBbV1HfAAllyuABbMRhB8wABGqFAERp4x5gA00UZBgQHwRvzIAGb++Mh8injUYiDlpUUA7sAszHJakjK29rp68ABEOXz5hcVdQA)**

### LINKED ISSUE #54054: The child class should not be able to access the class field defined by the parent class via super
# Bug Report

The child class should not be able to access the class field defined by the parent class via super

### 🔎 Search Terms

class field super parent

### 🕗 Version & Regression Information

Happens on 5.1.0-beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.0-beta#code/MYGwhgzhAEAq0G8BQ1XQGYEsCmIAm0AvNABQCURAfIgL5J2iQywBM02AHgC7YB2ezRCjTpyQtBOgQArgAdsAJwB0WXHnLDUdOkl7YA7nBbkV5IA)

### 💻 Code

```ts
class T {
    field = () => {}
}
class T2 extends T {
    f() {
        super.field()
//      ~~~~~~~~~~~
    }
}

new T2().f()
```

### 🙁 Actual behavior

No type error

### 🙂 Expected behavior

Type error.

`super` will look for the parent class, but the class field is always defined on the current instance.

### LINKED ISSUE #54900: `super` keyword allows calling arrow functions defined on the super class, but shouldn't
# Bug Report

### 🔎 Search Terms

super should not expose or allow call to superclass instance arrow function in subclass

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `super`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.3#code/MYGwhgzhAECCB2BLAtmE0DeAoavqoGsBTAOQHtEIjoBeaACgEpaA+aYM+CMkIgOhBkA5vQDkYJKnTwKVCKMZYAvliyhIMACLDoRAB4AXIvAAmMBCjSYceQqVnU6TVtAgBXAA5EATnzvlKIiZlVQ4uA2gTWmh4IgB3aG0RRSwTPzBiAKomIA)

### 💻 Code

```ts
class Animal {
    makeNoise = () => console.log('animal noises')
}

class Dog extends Animal {
    makeNoise = () => super.makeNoise()
}

const d = new Dog()

d.makeNoise()
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class Animal {
    constructor() {
        this.makeNoise = () => console.log('animal noises');
    }
}
class Dog extends Animal {
    constructor() {
        super(...arguments);
        this.makeNoise = () => super.makeNoise();
    }
}
const d = new Dog();
d.makeNoise();

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=5.1.3#code/MYGwhgzhAECCB2BLAtmE0DeAoavqoGsBTAOQHtEIjoBeaACgEpaA+aYM+CMkIgOhBkA5vQDkYJKnTwKVCKMZYAvliyhIMACLDoRAB4AXIvAAmMBCjSYceQqVnU6TVtAgBXAA5EATnzvlKIiZlVQ4uA2gTWmh4IgB3aG0RRSwTPzBiAKomIA)
      

### 🙁 Actual behavior

The TS compiler does not give any warnings or errors, and this code crashes at runtime.

```
[ERR]: "Executed JavaScript Failed:" 
[ERR]: super.makeNoise is not a function 
```

### 🙂 Expected behavior

The TS compiler should not allow me to call `super.makeNoise` if `Animal`'s `makeNoise` is an instance property/function, since only class methods will be on the prototype that `super` will access.

TS docs on why `super` doesn't work in practice here: https://www.typescriptlang.org/docs/handbook/2/classes.html#arrow-functions

### A note on why I care

Obviously the example is oversimplified. How I got here: I am trying to get contravariant type checking on subclass overrides of abstract super class functions, and as discussed in several places (below are some examples), you can only achieve this with superclass abstract instance properties that are functions (and not with abstract class methods):

- https://stackoverflow.com/a/75987382/2544629
- https://stackoverflow.com/a/45882417/2544629

But I would also like to let these subclasses sometimes call a superclass's default implementation (i.e. make some of the superclass methods not `abstract` but still overridable, and let them call `super.foo()` to defer to the default implementation if they choose to along with other logic). 

It seems like I am stuck wanting two things, when I'm only allowed to have one: either contravariant type checking via instance property functions that can't be referred to by the subclass via `super`, or overridable-but-defined superclass class methods whose subclass overrides are bivariantly checked.

I decided that contravariant type checking was more important to me, and I decided I'd get rid of the `super` calls everywhere -I found several type errors/small bugs in our codebase when I switched things over to instance property functions (mostly copy/paste errors or mistakes in generics on the class definition that weren't being caught in a bivariant world), and I was excited to try and get rid of the possibility for these mistakes going forward. 

However, I am concerned that the compiler won't stop my coworkers (or my future self) from exercising muscle memory in calling `super` in no-longer-method subclass overrides of instance property functions, which we've been doing for some time. Even if I find them all without the compiler's help now, the bug I'm reporting here prevents me from having any protection against regressions!

### Related issues

None of these are quite on the same topic, but they are in the same realm.

- https://github.com/microsoft/TypeScript/pull/18654
- https://github.com/microsoft/TypeScript/issues/51261
- https://github.com/microsoft/TypeScript/issues/27965

#### Possible duplicates mentioned below

- https://github.com/microsoft/TypeScript/issues/54054
- https://github.com/microsoft/TypeScript/issues/35314

#### Possible fix mentioned below

- https://github.com/microsoft/TypeScript/pull/54056

### PR DESCRIPTION
**Pull request by Jack-Works**
fixes #54054
fixes #35314
fixes #54900

### CODE REVIEW

**Comment by Guergeiro** [comment]
Just a question, does this handle the JS getter methods? Get methods work with super call in JS.

**Comment by Jack-Works** [comment]
> Just a question, does this handle the JS getter methods? Get methods work with super call in JS.

Only class fields are affected. getter and setters work normally.

File: tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+==== esDecorators-classExpression-classSuper.2.ts (1 errors) ====
+    declare var dec: any;
+    
+    // class expression in extends should not get an assigned name
+    (@dec
+    class C1 extends class { } {
+        static {
+            super.name;
+                  ~~~~
+!!! error TS2850: Class field 'name' defined by the parent class is not accessible in the child class via super.
```
Comment (andrewbranch) [inline tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt:13]: I don’t think this looks right. If we do

```ts
class C1 extends class Expr {} {
    static {
        console.log(super.name);
    }
}
```

this successfully logs `"Expr"`. Another example:

```ts
class C1 extends Number {
    static {
        console.log(super.EPSILON);
    }
}
```

It seems like “is a property declaration in an interface” does not imply “is implemented as a class field” in practice (particularly for built-in type constructors, but I also worry about DefinitelyTyped and other hand-written typings. I think we need to take the more conservative approach of looking for an actual class field declaration.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt (hunk 1 above)
Comment (Jack-Works) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt:13]: I can exclude the static initialization block for now.

Comment (andrewbranch) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt:13]: It’s not specific to the static case; this executes fine as well:

```ts
class C1 extends Array {
    foo() {
        console.log(super.length);
    }
}

new C1().foo()
```

A bad idea? Probably. But all these built-ins are not classes with class fields; they’re more like old-school constructor functions. So I think we should never issue this error unless we literally see class field syntax.

Comment (Jack-Works) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt:13]: > It’s not specific to the static case; this executes fine as well:
> 
> ```ts
> class C1 extends Array {
>     foo() {
>         console.log(super.length);
>     }
> }
> 
> new C1().foo()
> ```
> 
> A bad idea? Probably. But all these built-ins are not classes with class fields; they’re more like old-school constructor functions. So I think we should never issue this error unless we literally see class field syntax.

Oh... can we trust the class field in a declaration file?

Comment (Jack-Works) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt]: I changed it to check its declaration, I think it covers most false positives.

Comment (andrewbranch) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt]: Yeah, I think a class field in a declaration file is fair game. I’ll take a look at the changes.

Comment (andrewbranch) [reply tests/baselines/reference/esDecorators-classExpression-classSuper.2.errors.txt]: I think this looks good; can you add my examples to the tests?

**Comment by ekilah** [comment]
I was referred to this fix from my issue:
- https://github.com/microsoft/TypeScript/issues/54900

**Review by andrewbranch — APPROVED** [review]

**Comment by andrewbranch** [comment]
@typescript-bot run dt
@typescript-bot test top100

**Comment by andrewbranch** [comment]
[These errors](https://github.com/microsoft/TypeScript/pull/54056#issuecomment-1625923868) might be interesting to look at.

**Comment by andrewbranch** [comment]
In appsmith, the code presumably works at runtime because the compile target is `es6` where class fields aren’t supported. VS Code uses `useDefineForClassFields: false` which I assume has the same effect for this context. We could choose to only issue this error when class fields will be untransformed, but (1) the code is wrong even if it will technically work, and (2) we don’t know anything about the emit of dependency code, so it would be better to do this consistently. Since we’re past 5.2-beta, I think we should wait until 5.3 to merge this breaking change.

**Comment by Jack-Works** [comment]
> In appsmith, the code presumably works at runtime because the compile target is `es6` where class fields aren’t supported. VS Code uses `useDefineForClassFields: false` which I assume has the same effect for this context. We could choose to only issue this error when class fields will be untransformed, but (1) the code is wrong even if it will technically work, and (2) we don’t know anything about the emit of dependency code, so it would be better to do this consistently. Since we’re past 5.2-beta, I think we should wait until 5.3 to merge this breaking change.

Sorry, I'm confused, in what case it can be valid? I cannot think of a setup that defines a class field on the prototype instead of the instance.

**Comment by andrewbranch** [comment]
Hm, I assumed that a quirk of the transformation down to ES2015 would make it work somehow, but I think you’re right. Maybe this stuff isn’t working at all? This is very odd 🤔

File: tests/cases/compiler/classFieldSuperNotAccessible.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// @target: esnext
+class T {
+    field = () => {}
+}
+class T2 extends T {
+    f() {
+        super.field()
+        //    ~~~~~ error here
```
Comment (sandersn) [inline tests/cases/compiler/classFieldSuperNotAccessible.ts:8]: Copying the error span into the test is actually a little confusing when reading the error baseline. Maybe just a single caret?

**Comment by sandersn** [comment]
Even stranger, the first error, in appsmith, looks like a typo.

**Review by sandersn — APPROVED** [review]
I agree that this should go into 5.3 to avoid breaking vscode after the 5.2 beta.

In the meantime, I had one small nit about the test, but I'm not 100% what a better solution is.

**Pull request merged**